src/Sarif/ExtensionMethods.cs

```diff
@@ -253,86 +254,26 @@ public static string GetFirstSentence(string text)
+            string[] lineBreaks = { Environment.NewLine, "\r", "\n" };
 
-            foreach (char ch in text)
+            foreach (string s in lineBreaks)
             {
-                length++;
-                switch (ch)
-                {
-                    case '\'':
-                    {
-                        // we'll ignore everything within parenthized text
-                        if (!withinParentheses)
-                        {
-                            withinQuotes = !withinQuotes;
-                        }
-                        lastEncounteredWasDot = false;
-                        break;
-                    }
+                int index = text.IndexOf(s);
 
-                    case '(':
-                    {
-                        if (!withinQuotes)
-                        {
-                            withinParentheses = true;
-                        }
-                        lastEncounteredWasDot = false;
-                        break;
-                    }
-
-                    case ')':
-                    {
-                        if (!withinQuotes)
-                        {
-                            withinParentheses = false;
-                        }
-                        lastEncounteredWasDot = false;
-                        break;
-                    }
-
-                    case '.':
-                    {
-                        if (withinQuotes || withinParentheses || withinEllipsis) { continue; }
-                        if (length < text.Length && text[length] == '.')
-                        {
-                            withinEllipsis = true;
-                            lastEncounteredWasDot = false;
-                            break;
-                        }
-
-                        lastEncounteredWasDot = true;
-                        break;
-                    }
-
-                    // If we encounter a line-break, we return all leading text.
-                    case '\n':
-                    case '\r':
-                    {
-                        if (withinQuotes || withinParentheses) { continue; }
-                        return text.Substring(0, length).TrimEnd('\r', '\n', ' ', '.') + ".";
-                    }
+                if (index > -1)
+                {
+                    text = text.Substring(0, index);
+                }
+            }
 
-                    // If we encounter a space following a period, return 
-                    // all text terminating in the period (inclusive).
-                    case ' ':
-                    {
-                        if (!lastEncounteredWasDot) continue;
-                        if (withinQuotes || withinParentheses) { continue; }
-                        return text.Substring(0, length).TrimEnd('\r', '\n', ' ', '.') + ".";
-                    }
+            Match match = Regex.Match(text, @"^.*?[.?!][\)""']*(?=\s+\p{P}*[\p{Lu}\p{N}]|\s*$)", RegexOptions.Compiled);
```

Comment:
>@"^.*?[.?!][\)""']*(?=\s+\p{P}*[\p{Lu}\p{N}]|\s*$ [](start = 44, length = 49)

Oh good heavens :-)

I suggest you also use [`RegexOptions.IgnorePatternWhitespace`](https://docs.microsoft.com/en-us/dotnet/api/system.text.regularexpressions.regexoptions?view=netframework-4.7.2#System_Text_RegularExpressions_RegexOptions_IgnorePatternWhitespace), which allows you to document your pattern inline, for example:
```
@"^    # Beginning of string
.*?      # Any characters <-- and by the way, you don't need the ? because .* already means "0 or more"
[.?!]   # Sentence-ending punctuation
[\)""']*   # Zero or more quotation marks <-- and by the way (1) Why is the paren character in this character class, and (2) why is it escaped? You don't have to escape anything in a character class except you have to be careful with ^ and ]
(?=...    # Good heavens, I don't know what happens from here on out :-) :)
```

Reply:
Ok, I understand why your character class `[\)""']` includes the closing paren: because you want to match `(This was a test.)` as a first sentence.

But no, it [doesn't need to be escaped](https://www.regular-expressions.info/charclass.html):

> The usual metacharacters are normal characters inside a character class, and do not need to be escaped by a backslash. To search for a star or plus, use [+*]. Your regex will work fine if you escape the regular metacharacters inside a character class, but doing so significantly reduces readability.

---
In reply to: [213076317](https://github.com/Microsoft/sarif-sdk/pull/1050#discussion_r213076317) [](ancestors = 213076317)

Reply:
Correct re: paren et. al.
Removed the escape, remnant from earlier generation of the pattern.

---
In reply to: [213100457](https://github.com/Microsoft/sarif-sdk/pull/1050#discussion_r213100457) [](ancestors = 213100457,213076317)

Reply:
Nice documentation.

---
In reply to: [213135448](https://github.com/Microsoft/sarif-sdk/pull/1050#discussion_r213135448) [](ancestors = 213135448,213100457,213076317)